test/integrationtests/voight_kampff/tools.py

```diff
@@ -121,18 +156,46 @@ def wait_for_dialog(bus, dialogs, context=None, timeout=None):
                        provided by context or 10 seconds
     """
     if context:
-        timeout = timeout or context.step_timeout
+        timeout_duration = timeout or context.step_timeout
     else:
-        timeout = timeout or TIMEOUT
-    start_time = time.monotonic()
-    while time.monotonic() < start_time + timeout:
+        timeout_duration = timeout or DEFAULT_TIMEOUT
+    wait_for_dialog_match(bus, dialogs, timeout_duration)
+
+
+def wait_for_dialog_match(bus, dialogs, timeout=DEFAULT_TIMEOUT):
+    """Match dialogs spoken to the specified list of expected dialogs.
+
+    Only one of the dialogs in the provided list need to match for this
+    check to be successful.
+
+    Args:
+        bus (InterceptAllBusClient): Bus instance to listen on
+        dialogs (list): list of acceptable dialogs
+        timeout (int): how long to wait for the message, defaults to timeout
+                       provided by context or 10 seconds
+
+    Returns:
+        A boolean indicating if a match was found and the list of "speak"
+        events found on the message bus during the matching process.
+    """
+    match_found = False
+    speak_messages = list()
+    timeout_time = time.monotonic() + timeout
+    while time.monotonic() < timeout_time:
         for message in bus.get_messages('speak'):
+            speak_messages.append(message)
             dialog = message.data.get('meta', {}).get('dialog')
+            print('dialog: ', dialog, '\tutterance: ', message.data['utterance'])
             if dialog in dialogs:
-                bus.clear_messages()
-                return
-        bus.new_message_available.wait(0.5)
-    bus.clear_messages()
+                wait_while_speaking()
+                match_found = True
+                break
+        bus.clear_messages()
+        if match_found:
+            break
+        time.sleep(1)
```

Comment:
can `bus.new_message_available.wait(0.5)` be used to make it a bit more snappy?

Reply:
In my testing, using the message_available event caused a very tight loop because there are so many message bus messages being emitted.  Instead of looping ten times, as this code does, it looped hundreds of times.  I would like to avoid these types of waits completely and just use message bus events with timeouts.

Reply:
Not seeing the issue with the tight loop, this waits for messages as soon as messages arrive we want to evaluate them. But you are right in this case an event driven approach would be much tidier. Can we change the implementation to something similar to:

```python
def wait_for_dialog_match(bus, dialogs, timeout=DEFAULT_TIMEOUT):
    """Match dialogs spoken to the specified list of expected dialogs.
    Only one of the dialogs in the provided list need to match for this
    check to be successful.
    Args:
        bus (InterceptAllBusClient): Bus instance to listen on
        dialogs (list): list of acceptable dialogs
        timeout (int): how long to wait for the message, defaults to timeout
                       provided by context or 10 seconds
    Returns:
        A boolean indicating if a match was found and the list of "speak"
        events found on the message bus during the matching process.
    """
    # Check if the dialog has already been received
    historical_result = check_historical_messages()
    if historical_result:
        return historical_result

    match_found = Event()
    speak_messages = list()

    def dialog_handler(message):
        nonlocal match_found
        nonlocal speak_messages
        dialog = message.data.get('meta', {}).get('dialog')
        speak_messages.append(message)
        if dialog in dialogs:
            wait_while_speaking()
            match_found.set()

    bus.on('speak', dialog_handler)
    match_found(timeout=timeout)
    bus.remove('speak', dialog_handler)
    return match_found.is_set(), speak_messages
```